src/room_member.cpp

```diff
@@ -0,0 +1,196 @@
+    stream.Write(static_cast<uint8_t>(wifi_packet.type));
+    stream.Write(wifi_packet.channel);
+    stream.Write(wifi_packet.transmitter_address);
+    stream.Write(wifi_packet.destination_address);
+    stream.Write(static_cast<uint32_t>(wifi_packet.data.size()));
+    stream.Write(reinterpret_cast<char*>(wifi_packet.data.data()), wifi_packet.data.size());
+
+    peer->Send(&stream, HIGH_PRIORITY, RELIABLE_ORDERED, 0, server_address, false);
+}
+
+/**
+ * Sends a request to the server, asking for permission to join a room with the specified nickname and preferred mac.
+ * @params nickname The desired nickname.
+ * @params preferred_mac The preferred MAC address to use in the room, an all-zero MAC address tells the server to assign one for us.
+ */
+static void SendJoinRequest(RakNet::RakPeerInterface* peer, const std::string& nickname, const MacAddress& preferred_mac = NoPreferredMac) {
+    // TODO(Subv): Implement.
+}
+
+void RoomMember::ReceiveLoop() {
+    // Receive packets while the connection is open
+    while (IsConnected()) {
+        RakNet::Packet* packet = peer->Receive();
+        if (!packet)
+            continue;
+
+        switch (packet->data[0]) {
+        case ID_ROOM_CHAT:
+            //FIXME
+            break;
+        case ID_ROOM_WIFI_PACKET:
+            HandleWifiPackets(packet);
+            break;
+        case ID_ROOM_INFORMATION:
+            //FIXME: Update playerlist from packet
+            break;
+        case ID_ROOM_NAME_COLLISION:
+            state = State::NameCollision;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_ROOM_MAC_COLLISION:
+            state = State::MacCollision;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_ROOM_JOIN_SUCCESS:
+            // The join request was successful, we are now in the room.
+            // If we joined successfully, there must be at least one client in the room: us.
+            ASSERT_MSG(GetMemberInformation().size() > 0, "We have not yet received member information.");
+            state = State::Joined;
+            break;
+        case ID_DISCONNECTION_NOTIFICATION:
+            // Connection lost normally
+            state = State::Idle;
+            peer->CloseConnection(server_address, true);
+            return;
+        case ID_INCOMPATIBLE_PROTOCOL_VERSION:
+            state = State::WrongVersion;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_CONNECTION_ATTEMPT_FAILED:
+            state = State::Error;
+            break;
+        case ID_NO_FREE_INCOMING_CONNECTIONS:
+            state = State::RoomFull;
+            peer->CloseConnection(server_address, true);
+            break;
+        case ID_CONNECTION_LOST:
+            // Couldn't deliver a reliable packet, the other system was abnormally terminated
+            state = State::LostConnection;
+            break;
```

Comment:
Did you check which events are automaticly closing the connection and which don't?
If the server is full I'd expect raknet to internally have closed the connection. if it doesn't it might not do so here either? (maybe it will just try to reconnect when continuing sending data?)